tagmanager/src/tm_workspace.c

```diff
@@ -823,514 +728,333 @@ fill_find_tags_array (GPtrArray *dst, const GPtrArray *src,
+	gboolean member, const gchar *current_scope)
+{
+	GPtrArray *member_tags = NULL;
+	guint i;
 
-		if ((tags2) && (tags2->len == 1) && (tag = TM_TAG (tags2->pdata[0])))
+	/* there may be several variables/types with the same name - try each of them until
+	 * we find something */
+	for (i = 0; i < tags->len; i++)
+	{
+		TMTag *tag = TM_TAG(tags->pdata[i]);
+		TMTagType types = (tm_tag_class_t | /* tm_tag_namespace_t | tm_tag_package_t | */
+						   tm_tag_struct_t | tm_tag_union_t |
+						   tm_tag_enum_t | tm_tag_typedef_t);
+
+		if (tag->type & types)  /* type: namespace search */
+			member_tags = find_scope_members(searched_array, tag->name, lang, TRUE);
+		else if (tag->var_type)  /* variable: scope search */
 		{
-			if (tag->type == tm_tag_typedef_t && tag->var_type
-				&& tag->var_type[0] != '\0')
-			{
-				new_name = tag->var_type;
-				continue;
-			}
-			filename = (tag->file ?
-						tag->file->short_name : NULL);
-			if (tag->scope && tag->scope[0] != '\0')
+			if (!(tag->type & (tm_tag_field_t | tm_tag_member_t)) || member ||
+				member_at_method_scope(current_scope, tag->scope, lang))
 			{
-				del = 1;
-				if (tag->file &&
-					tag->file->lang == langJava)
-				{
-					new_name = g_strdup_printf ("%s.%s",
-												tag->scope,
-												new_name);
-				}
-				else
-				{
-					new_name = g_strdup_printf ("%s::%s",
-												tag->scope,
-												new_name);
-				}
+				gchar *tag_type = g_strdup(tag->var_type);
+
+				/* remove pointers in case the type contains them */
+				g_strdelimit(tag_type, "*", ' ');
+				g_strstrip(tag_type);
+				member_tags = find_scope_members(searched_array, tag_type, lang, FALSE);
+				g_free(tag_type);
 			}
-			break;
-		}
-		else
-		{
-			return NULL;
 		}
+
+		if (member_tags)
+			break;
 	}
 
-	g_ptr_array_set_size (tags, 0);
+	return member_tags;
+}
 
-	if (tag && tag->file)
-	{
-		local = tm_tags_extract (tag->file->tags_array,
-								 (tm_tag_function_t |
-								  tm_tag_field_t | tm_tag_enumerator_t |
-								  tm_tag_namespace_t | tm_tag_class_t ));
-	}
-	else
-	{
-		local = tm_tags_extract (theWorkspace->tags_array,
-								 (tm_tag_function_t | tm_tag_prototype_t |
-								  tm_tag_member_t |
-								  tm_tag_field_t | tm_tag_enumerator_t |
-								  tm_tag_namespace_t | tm_tag_class_t ));
-	}
 
-	if (local)
-	{
-		found = find_namespace_members_tags (local, tags,
-										 langJava, new_name, filename);
-		g_ptr_array_free (local, TRUE);
-	}
+/* Returns all member tags of a struct/union/class if the provided name is a variable
+ of of such a type or the name of the type.
+ @param source_file TMSourceFile of the edited source file or NULL if not available
+ @param name Name of the variable/type whose members are searched
+ @param function TRUE if the name is a name of a function
+ @param member TRUE if invoked on class/struct member (e.g. after the last dot in foo.bar.)
+ @param current_scope The current scope in the editor
```

Comment:
No - it's the scope where you are editing and not the scope at which the variable/type was defined.